File: ios/Runner/RecordingPlugin.swift
import Foundation
import Flutter
import AVFoundation
import Photos

class RecordingPlugin: NSObject, FlutterPlugin {
    private var methodChannel: FlutterMethodChannel?
    private var eventChannel: FlutterEventChannel?
    private var eventSink: FlutterEventSink?
    private var isRecording = false
    private var currentRecordingId: String?
    private var outputFile: URL?
    
    static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "synqcast_recording", binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(name: "synqcast_recording_events", binaryMessenger: registrar.messenger())
        let instance = RecordingPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        eventChannel.setStreamHandler(instance)
        instance.methodChannel = channel
        instance.eventChannel = eventChannel
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "startRecording":
            startRecording(call: call, result: result)
        case "stopRecording":
            stopRecording(result: result)
        case "getRecordings":
            getRecordings(result: result)
        case "deleteRecording":
            deleteRecording(call: call, result: result)
        case "shareRecording":
            shareRecording(call: call, result: result)
        case "requestRecordingPermission":
            requestPermission(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func startRecording(call: FlutterMethodCall, result: @escaping FlutterResult) {
        if isRecording {
            result(FlutterError(code: "ALREADY_RECORDING", message: "Recording is already active", details: nil))
            return
        }
        
        guard let args = call.arguments as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments", details: nil))
            return
        }
        
        let recordingId = args["recordingId"] as? String ?? "recording_\(Date().timeIntervalSince1970)"
        let roomName = args["roomName"] as? String
        let width = args["width"] as? Int ?? 1920
        let height = args["height"] as? Int ?? 1080
        let frameRate = args["frameRate"] as? Int ?? 30
        let bitRate = args["bitrate"] as? Int ?? 6000000
        let includeAudio = args["includeAudio"] as? Bool ?? true
        
        // Create output file
        let documentsPath = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let fileName = "synqcast_\(roomName ?? "recording")_\(Date().timeIntervalSince1970).mp4"
        outputFile = documentsPath.appendingPathComponent(fileName)
        
        guard let outputFile = outputFile else {
            result(FlutterError(code: "FILE_ERROR", message: "Failed to create output file", details: nil))
            return
        }
        
        // Check permissions
        AVCaptureDevice.requestAccess(for: .video) { [weak self] granted in
            DispatchQueue.main.async {
                if granted {
                    self?.startAVRecording(outputFile: outputFile, recordingId: recordingId, result: result)
                } else {
                    result(FlutterError(code: "PERMISSION_DENIED", message: "Camera permission denied", details: nil))
                }
            }
        }
    }
    
    private func startAVRecording(outputFile: URL, recordingId: String, result: @escaping FlutterResult) {
        // For iOS, we'll use a simplified approach since full screen recording requires ReplayKit
        // This is a placeholder implementation that would need to be expanded for full functionality
        
        isRecording = true
        currentRecordingId = recordingId
        
        // Send event
        sendEvent(type: "recording_started", data: nil)
        
        result([
            "success": true,
            "filePath": outputFile.path
        ])
    }
    
    private func stopRecording(result: @escaping FlutterResult) {
        if !isRecording {
            result(["success": true])
            return
        }
        
        isRecording = false
        currentRecordingId = nil
        
        // Send event
        sendEvent(type: "recording_stopped", data: nil)
        
        result(["success": true])
    }
    
    private func getRecordings(result: @escaping FlutterResult) {
        let documentsPath = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        var recordings: [[String: Any]] = []
        
        do {
            let files = try FileManager.default.contentsOfDirectory(at: documentsPath, includingPropertiesForKeys: nil)
            for file in files {
                if file.lastPathComponent.hasPrefix("synqcast_") && file.pathExtension == "mp4" {
                    let attributes = try FileManager.default.attributesOfItem(atPath: file.path)
                    let fileSize = attributes[.size] as? Int64 ?? 0
                    
                    let recording: [String: Any] = [
                        "filename": file.lastPathComponent,
                        "filePath": file.path,
                        "size": fileSize,
                        "startTime": nil,
                        "endTime": nil,
                        "duration": 0,
                        "roomName": file.lastPathComponent.replacingOccurrences(of: "synqcast_", with: "").replacingOccurrences(of: ".mp4", with: "")
                    ]
                    recordings.append(recording)
                }
            }
        } catch {
            print("Error getting recordings: \(error)")
        }
        
        result([
            "success": true,
            "recordings": recordings
        ])
    }
    
    private func deleteRecording(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let filename = args["filename"] as? String else {
            result(FlutterError(code: "INVALID_FILENAME", message: "Filename is required", details: nil))
            return
        }
        
        let documentsPath = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let fileURL = documentsPath.appendingPathComponent(filename)
        
        do {
            try FileManager.default.removeItem(at: fileURL)
            result(["success": true])
        } catch {
            result(FlutterError(code: "DELETE_ERROR", message: "Failed to delete recording", details: error.localizedDescription))
        }
    }
    
    private func shareRecording(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let filename = args["filename"] as? String else {
            result(FlutterError(code: "INVALID_FILENAME", message: "Filename is required", details: nil))
            return
        }
        
        let documentsPath = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let fileURL = documentsPath.appendingPathComponent(filename)
        
        if FileManager.default.fileExists(atPath: fileURL.path) {
            // In a real implementation, this would use UIActivityViewController to share the file
            // For now, we'll just return success
            result(["success": true])
        } else {
            result(FlutterError(code: "FILE_NOT_FOUND", message: "Recording file not found", details: nil))
        }
    }
    
    private func requestPermission(result: @escaping FlutterResult) {
        AVCaptureDevice.requestAccess(for: .video) { granted in
            DispatchQueue.main.async {
                result(["granted": granted])
            }
        }
    }
    
    private func sendEvent(type: String, data: [String: Any]?) {
        let event: [String: Any] = [
            "type": type,
            "data": data ?? [:]
        ]
        eventSink?(event)
    }
}

// MARK: - FlutterStreamHandler

extension RecordingPlugin: FlutterStreamHandler {
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}

File: android/app/src/main/kotlin/com/synqcast/synqcast/RecordingPlugin.kt
package com.synqcast.synqcast

import android.content.Context
import android.media.MediaRecorder
import android.os.Build
import android.util.Log
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.io.File

class RecordingPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    
    private var mediaRecorder: MediaRecorder? = null
    private var isRecording = false
    private var currentRecordingId: String? = null
    private var outputFile: File? = null
    
    companion object {
        private const val TAG = "RecordingPlugin"
        private const val CHANNEL = "synqcast_recording"
    }

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, CHANNEL)
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        // Clean up resources without result callback
        if (isRecording) {
            try {
                mediaRecorder?.apply {
                    stop()
                    release()
                }
                mediaRecorder = null
                isRecording = false
                currentRecordingId = null
            } catch (e: Exception) {
                Log.e(TAG, "Error cleaning up recording", e)
            }
        }
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "startRecording" -> startRecording(call, result)
            "stopRecording" -> stopRecording(result)
            "getRecordings" -> getRecordings(result)
            "deleteRecording" -> deleteRecording(call, result)
            "shareRecording" -> shareRecording(call, result)
            "requestRecordingPermission" -> requestPermission(result)
            else -> result.notImplemented()
        }
    }

    private fun startRecording(call: MethodCall, result: Result) {
        if (isRecording) {
            result.error("ALREADY_RECORDING", "Recording is already active", null)
            return
        }

        try {
            val recordingId = call.argument<String>("recordingId") ?: "recording_${System.currentTimeMillis()}"
            val roomName = call.argument<String>("roomName")
            val width = call.argument<Int>("width") ?: 1920
            val height = call.argument<Int>("height") ?: 1080
            val frameRate = call.argument<Int>("frameRate") ?: 30
            val bitRate = call.argument<Int>("bitrate") ?: 6000000
            val includeAudio = call.argument<Boolean>("includeAudio") ?: true

            // Create output file
            val fileName = "synqcast_${roomName ?: "recording"}_${System.currentTimeMillis()}.mp4"
            outputFile = File(context.getExternalFilesDir(null), fileName)
            
            // Initialize MediaRecorder
            mediaRecorder = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                MediaRecorder(context)
            } else {
                @Suppress("DEPRECATION")
                MediaRecorder()
            }

            mediaRecorder?.apply {
                setVideoSource(MediaRecorder.VideoSource.SURFACE)
                if (includeAudio) {
                    setAudioSource(MediaRecorder.AudioSource.MIC)
                }
                setOutputFormat(MediaRecorder.OutputFormat.MPEG_4)
                setVideoEncoder(MediaRecorder.VideoEncoder.H264)
                if (includeAudio) {
                    setAudioEncoder(MediaRecorder.AudioEncoder.AAC)
                }
                setVideoEncodingBitRate(bitRate)
                setVideoFrameRate(frameRate)
                setVideoSize(width, height)
                setOutputFile(outputFile?.absolutePath)
                
                prepare()
                start()
            }

            isRecording = true
            currentRecordingId = recordingId

            result.success(mapOf(
                "success" to true,
                "filePath" to outputFile?.absolutePath
            ))

        } catch (e: Exception) {
            Log.e(TAG, "Error starting recording", e)
            result.error("START_ERROR", "Failed to start recording", e.message)
        }
    }

    private fun stopRecording(result: Result) {
        if (!isRecording) {
            result.success(mapOf("success" to true))
            return
        }

        try {
            mediaRecorder?.apply {
                stop()
                release()
            }
            mediaRecorder = null
            isRecording = false
            currentRecordingId = null

            result.success(mapOf("success" to true))
        } catch (e: Exception) {
            Log.e(TAG, "Error stopping recording", e)
            result.error("STOP_ERROR", "Failed to stop recording", e.message)
        }
    }

    private fun getRecordings(result: Result) {
        try {
            val recordingsDir = context.getExternalFilesDir(null)
            val recordings = mutableListOf<Map<String, Any>>()
            
            recordingsDir?.listFiles()?.forEach { file ->
                if (file.name.startsWith("synqcast_") && file.extension in listOf("mp4", "webm")) {
                    val metadata = mapOf<String, Any>(
                        "filename" to file.name,
                        "filePath" to file.absolutePath,
                        "size" to file.length(),
                        "startTime" to "", // Would need to track this separately
                        "endTime" to "",
                        "duration" to 0, // Would need to extract from file
                        "roomName" to file.name.substringAfter("synqcast_").substringBefore("_")
                    )
                    recordings.add(metadata)
                }
            }

            result.success(mapOf(
                "success" to true,
                "recordings" to recordings
            ))
        } catch (e: Exception) {
            Log.e(TAG, "Error getting recordings", e)
            result.error("GET_ERROR", "Failed to get recordings", e.message)
        }
    }

    private fun deleteRecording(call: MethodCall, result: Result) {
        try {
            val filename = call.argument<String>("filename")
            if (filename == null) {
                result.error("INVALID_FILENAME", "Filename is required", null)
                return
            }

            val recordingsDir = context.getExternalFilesDir(null)
            val file = File(recordingsDir, filename)
            
            if (file.exists() && file.delete()) {
                result.success(mapOf("success" to true))
            } else {
                result.error("DELETE_ERROR", "Failed to delete recording", null)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error deleting recording", e)
            result.error("DELETE_ERROR", "Failed to delete recording", e.message)
        }
    }

    private fun shareRecording(call: MethodCall, result: Result) {
        try {
            val filename = call.argument<String>("filename")
            if (filename == null) {
                result.error("INVALID_FILENAME", "Filename is required", null)
                return
            }

            val recordingsDir = context.getExternalFilesDir(null)
            val file = File(recordingsDir, filename)
            
            if (file.exists()) {
                // In a real implementation, this would use Intent to share the file
                // For now, we'll just return success
                result.success(mapOf("success" to true))
            } else {
                result.error("FILE_NOT_FOUND", "Recording file not found", null)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error sharing recording", e)
            result.error("SHARE_ERROR", "Failed to share recording", e.message)
        }
    }

    private fun requestPermission(result: Result) {
        // For Android, recording permissions are typically handled at the app level
        // via the manifest file. This method can be used to check if permissions are granted.
        result.success(mapOf("granted" to true))
    }
}
